crates/matrix-sdk-base/src/store/indexeddb_store.rs

```diff
@@ -469,6 +488,185 @@ impl IndexeddbStore {
+                        end: timeline.end.clone(),
+                        start_position: usize::MAX / 2,
+                        end_position: usize::MAX / 2,
+                    }
+                };
+
+                if timeline.sync {
+                    let mut room_version = None;
+                    for event in timeline.events.iter().rev() {
+                        // Redact events already in store only on sync response
+                        if let Ok(AnySyncRoomEvent::Message(AnySyncMessageEvent::RoomRedaction(
+                            redaction,
+                        ))) = event.event.deserialize()
+                        {
+                            let redacts_key = (room_id, &redaction.redacts).encode();
+                            if let Some(position_key) =
+                                event_id_to_position_store.get_owned(redacts_key)?.await?
+                            {
+                                if let Some(mut full_event) = timeline_store
+                                    .get(&position_key)?
+                                    .await?
+                                    .map(|e| {
+                                        self.deserialize_event::<SyncRoomEvent>(e)
+                                            .map_err(StoreError::from)
+                                    })
+                                    .transpose()?
+                                {
+                                    let inner_event = full_event.event.deserialize()?;
+                                    if room_version.is_none() {
+                                        room_version = Some(room_infos
+                                                        .get(&room_key)?
+                                                        .await?
+                                                        .map(|r| self.deserialize_event::<RoomInfo>(r))
+                                                        .transpose()?
+                                                        .and_then(|info| {
+                                                            info.base_info
+                                                                .create
+                                                                .map(|event| event.room_version)
+                                                        }).unwrap_or_else(|| {
+                                                            warn!("Unable to find the room version for {}, assume version 9", room_id);
+                                                            RoomVersionId::V9
+                                                        }));
+                                    }
+
+                                    full_event.event = Raw::new(&AnySyncRoomEvent::from(
+                                        inner_event
+                                            .redact(redaction, room_version.as_ref().unwrap()),
+                                    ))?;
```

Comment:
the idea is that we only lookup the room version once per `TimelineSlice` and not for every single event we need to remove.

Reply:
we could also use `OnceCell` but i didn't want to introduce a new crate for this single use case.

Reply:
Well... but why within the loop though? Am I missing something or isn't everything within that call already available on line 592 and thus should just happen there outside/before the loop? Arguably as it happens in both cases of iteration - even outside of this?

Generally, more than 4 levels of indentation are very hard to read and follow (and makes the lines break more often than necessary), I wonder if we can't just make some of them simpler... like - the loops are the same in both cases, it is just the order of the list that differs, right? then the following should also reduce the code a lot:

```rust
let timeline_iter = if timeline.sync { timeline.events.iter().rev() } else { timeline.events.iter() };
```

(might needs a type annotation, haven't tested).

I'd appreciate if we could reduce the complexity of this code block...

Reply:
That was a bit too simplistic because they come back with different types. But if we used an `either` (probably better from `itertools`) this would be solved nicely. Example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=21ded866b7caa0917f3b16c11c35dc2d

Reply:
Before we check events we can't be sure if we will need the room version, therefore i look it up only at the first use